cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// The translation unit for reduction `minmax`
+
+#include <thrust/transform_reduce.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+/**
+ * @brief stores the minimum and maximum
+ * values that have been encountered so far
+ *
+ */
+template <typename T>
+struct minmax_pair {
+  T min_val;
+  T max_val;
+  bool min_valid;
+  bool max_valid;
+
+  __host__ __device__ minmax_pair()
+    : min_val(cudf::DeviceMin::identity<T>()),
+      max_val(cudf::DeviceMax::identity<T>()),
+      min_valid(false),
+      max_valid(false){};
+  __host__ __device__ minmax_pair(T val, bool valid_)
+    : min_val(val), max_val(val), min_valid(valid_), max_valid(valid_){};
+  __host__ __device__ minmax_pair(T min_val_, bool min_valid_, T max_val_, bool max_valid_)
+    : min_val(min_val_), max_val(max_val_), min_valid(min_valid_), max_valid(max_valid_){};
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Respects
+ * validity.
+ *
+ */
+template <typename T, bool has_nulls = true>
+struct minmax_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __device__ minmax_pair<T> operator()(minmax_pair<T> const &lhs, minmax_pair<T> const &rhs) const
+  {
+    T const x_min = (lhs.min_valid || !has_nulls) ? lhs.min_val : cudf::DeviceMin::identity<T>();
+    T const y_min = (rhs.min_valid || !has_nulls) ? rhs.min_val : cudf::DeviceMin::identity<T>();
+    T const x_max = (lhs.max_valid || !has_nulls) ? lhs.max_val : cudf::DeviceMax::identity<T>();
+    T const y_max = (rhs.max_valid || !has_nulls) ? rhs.max_val : cudf::DeviceMax::identity<T>();
```

Comment:
I just remembered that this code can be significantly simplified by using a null_replacement_iterator. See: https://github.com/rapidsai/cudf/blob/branch-0.16/cpp/include/cudf/detail/iterator.cuh#L106-L124

There's no need to keep track of the validity of the results in the `minmax_pair`. `null_replacement_iterator` will take care of injecting the identity value for null elements. The only thing you need to detect is if _all_ of the elements in the column are null.

Reply:
The whole thing can fit into existing reduction code except one. Problem is creating a pair of device_scalars and returning it. `cudf::scalar_type_t<OutputType>` for `OutputType=thrust::pair<T,T>` is not present.

With some changes to existing reduction code, it is possible to add `minmax` with few specialization. (compound_op could be modified to add a store_functor which can return a different datatype).

Reply:
I'm not sure I follow. The general run of this with a column of `5, 10, 15` is to generate a minmax pair from each value, so `minmax_pair(5, valid);`, etc. and then to start merging pairs inside the binary op. The result is that one invalid value would need the `DeviceMin::identity<T>()` and the `DeviceMax::identity<T>()` depending on if it represented a min or a max inside the minmax pair. The alternative would be to have branches around the comparison such as
```
if (p1.min_valid && p2.min_valid) {
  val = thrust::min(p1.min, p2.min);
} else if (p1.min_valid && !p2.min_valid) {
  val = p1.min;
} else if (!p1.min_valid && p2.min_valid) {
  val = p2.min;
} else {
  val = invalid_const_val;
}
```
This would allow us to ignore the invalid values entirely and then they could contain any value since it wasn't used for comparison, but I felt this was too branchy.

Since `make_null_replacement_iterator` can only accept a column as input and can only return a T that matches the column's T, I don't see how we can encode the information we need. If we had an iterator version, we could give it a transforming iterator that converted the column data into `minmax_pair<T>` and have the null replacement be `minmax_pair<T>{DeviceMin::identity(), DeviceMax::identity()}`, but that isn't possible currently. I assume because an iterator alone would lose the null information.

Maybe I don't understand where this iterator would go. My initial look I thought it would replace the counting iterator and we could just return a valid value for each entry and I was happy, but then I realized that the value needed to change depending on where it was being stuffed. This would lead to trying to infer if the entry was valid or not and cause it to be incorrect for actual max and min type values. Did you have a different way to do this in mind?

Reply:
Hm, you're right. My original idea doesn't work as I thought. I think we can still simplify it though using [`make_pair_iterator()`](https://github.com/rapidsai/cudf/blob/798267bd5d811b8441a2d2c875257637c59dc48d/cpp/include/cudf/detail/iterator.cuh#L154) that returns a `<T,bool>` for each element indicating the value of the element and it's validity.

Something like this:
```
column_device_view col;
auto pair = make_pair_iterator(col);
auto pair_to_min_max = make_transform_iterator(pair, [] __device__ (auto [e, v]){ return v ? minmax_pair{e,e} : minmax_pair{DeviceMin::idenity(), DeviceMax::identity()};
```